src/controllers/token.rs

```diff
@@ -114,3 +117,133 @@ pub fn revoke_current(req: &mut dyn RequestExt) -> EndpointResult {
+    let sig = base64::decode(sig)
+        .map_err(|e| bad_request(&format!("failed to decode signature as base64: {e:?}")))?;
+
+    // Fetch list of public keys from GitHub API
+    let app = req.app();
+    let public_keys = app
+        .github
+        .public_keys(&app.config.gh_client_id, &app.config.gh_client_secret)
+        .map_err(|e| bad_request(&format!("failed to fetch GitHub public keys: {e:?}")))?;
+
+    for key in public_keys {
+        if key.key_identifier == req_key_id {
+            if !key.is_current {
+                return Err(bad_request(&format!(
+                    "key id {req_key_id} is not a current key"
+                )));
+            }
+            let key_bytes =
+                key_from_spki(&key).map_err(|_| bad_request("cannot parse public key"))?;
+            let gh_key =
+                signature::UnparsedPublicKey::new(&signature::ECDSA_P256_SHA256_ASN1, &key_bytes);
+
+            return match gh_key.verify(json, &sig) {
+                Ok(v) => {
+                    info!(
+                        "GitHub secret alert request validated with key id {}",
+                        key.key_identifier
+                    );
+                    Ok(v)
+                }
+                Err(e) => Err(bad_request(&format!("invalid signature: {e:?}"))),
+            };
+        }
+    }
+
+    return Err(bad_request(&format!("unknown key id {req_key_id}")));
+}
+
+#[derive(Deserialize, Serialize)]
+struct GitHubSecretAlert {
+    token: String,
+    r#type: String,
+    url: String,
+    source: String,
+}
+
+/// Revokes an API token and notifies the token owner
+fn alert_revoke_token(
+    req: &dyn RequestExt,
+    alert: &GitHubSecretAlert,
+) -> Result<(), Box<dyn AppError>> {
+    let conn = req.db_write()?;
+    // not using ApiToken::find_by_api_token in order to preserve last_used_at
+    let token = api_tokens::table
+        .filter(api_tokens::token.eq(alert.token.as_bytes()))
+        .first::<ApiToken>(&*conn)?;
+
+    diesel::update(&token)
+        .set(api_tokens::revoked.eq(true))
+        .execute(&*conn)?;
+
+    // send email notification to the token owner
+    let user = User::find(&conn, token.user_id)?;
+    info!(
+        "Revoked API token '{}' for user {} ({})",
+        alert.token, user.gh_login, user.id
+    );
+    match user.email(&conn)? {
+        None => {
+            info!(
+                "No email address for user {} ({}), cannot send email notification",
+                user.gh_login, user.id
+            );
+            Ok(())
+        }
+        Some(email) => req.app().emails.send_token_exposed_notification(
+            &email,
+            &alert.url,
+            "GitHub",
+            &alert.source,
+            &token.name,
+        ),
+    }
+}
+
+/// Handles the `POST /tokens/alert/github` route.
+pub fn alert_github(req: &mut dyn RequestExt) -> EndpointResult {
```

Comment:
I'm wondering if we should move this endpoint and the related code into a dedicated module. something like `controllers::github::secret_scanning` maybe?

Reply:
Moved as suggested.